File: inf/lab2/main.py
import re

def main():
    while True:
        input_str = input("enter a message consisting of 7 digits 0 or 1, written without spaces:")
        if len(input_str) == 7 and re.fullmatch(r'[01]*', input_str):
            int_arr = [int(char) for char in input_str]
            break
        else:
            print("you must enter 7 bits containing only 0 and 1!")

    s1 = (int_arr[0] + int_arr[2] + int_arr[4] + int_arr[6]) % 2
    s2 = (int_arr[1] + int_arr[2] + int_arr[5] + int_arr[6]) % 2
    s3 = (int_arr[3] + int_arr[4] + int_arr[5] + int_arr[6]) % 2

    err_pos = s1 + s2 * 2 + s3 * 4

    bit_type = {
        0: "correct",
        1: "r1",
        2: "r2",
        3: "i1",
        4: "r3",
        5: "i2",
        6: "i3",
        7: "i4"
    }.get(err_pos, "unknown")

    print(bit_type)

    with open("result.txt", "w") as file:
        file.write(bit_type)

if __name__ == "__main__":
    main()

File: inf/lab2/src/main/kotlin/Main.kt
package ru.pozitp

import java.io.File

fun main() {
    // r1 r2 i1 r3 i2 i3 i4
    val intArr: MutableList<Int>
    while (true) {
        println("enter a message consisting of 7 digits 0 or 1, written without spaces:")
        val input = readln()
        if (input.length == 7 && Regex("[01]*").matches(input)) {
            intArr = input.map { it.toString().toInt() }.toMutableList()
            break
        } else {
            println("you must enter 7 bits containing only 0 and 1!")
        }
    }

    val s1 = (intArr[0] + intArr[2] + intArr[4] + intArr[6]) % 2
    val s2 = (intArr[1] + intArr[2] + intArr[5] + intArr[6]) % 2
    val s3 = (intArr[3] + intArr[4] + intArr[5] + intArr[6]) % 2

    val errPos = s1 + s2 * 2 + s3 * 4

    val bitType = when (errPos) {
        0 -> "correct"
        1 -> "r1"
        2 -> "r2"
        3 -> "i1"
        4 -> "r3"
        5 -> "i2"
        6 -> "i3"
        7 -> "i4"
        else -> "unknown"
    }
    println(bitType)

    File("result.txt").writeText(bitType)
}
